### PR TITLE
BoundsControl final migration step

### PR DESCRIPTION
**Pull request by thalbern**
## Overview
- upgraded all prefabs and scenes (except boundingbox example) to use bounds control
- added prefab support for migrationhandler as prefabs can't store inline scriptables and need to be added to an existing or stored as separate asset.
- adjusted how we're displaying scriptables in our ui util -> there's 5 different modes now on how we're displaying scriptables to the user - precise description of the modes can be found in the comments in the code.
- added warning to migration window to inform users that they might loose data for objects that have prefab property overrides.
- fixed wrong rotation calculation in rotation handles
- fixed wrong usage of transform passed into proximity effect update
- fixed wrong usage of transform for rotation calculations
- fixed wrong return value of scale min / max in bounding box
- adjusted some of the prefab values of cheese and slate -> cheese proximity effect and handle sizes values were set to crazy numbers / slate had 35 cm flatten axis display scale which wasn't noticeable in boundingbox because it was broken, but now works properly, so reset the number to 0 for now.
- moved deprecated warning + button helper function into migrationtool / changed to generic and added to bounding box inspector

## Changes
- Fixes: https://github.com/microsoft/MixedRealityToolkit-Unity/issues/7358 

### Scriptables in Inspector Visualization Cases

**GameObject is part of a Prefab and no scriptable has been linked**
image

**GameObject is part of a Prefab and scriptable is nested as a serialized object inside the prefab asset file**
image

**GameObject is an instance of a prefab and scriptable is part of the prefab file**
image

**GameObject is an instance in a scene and scriptable is a local nested instance which is part of the scene file**
image

**Scriptable property is referring to an external scriptable asset file.**
image

### Deprecated Warning in Bounding Box
image

### Migration Window Warning
image

## Verification

This change updates all prefabs and scenes using BoundingBox to BoundsControl. To verify have a look at the updated prefabs and scenes in the following lists.

**Scene list** (scenes that are either using those prefabs or having their own bounding boxes in gameobject instances)

- HandInteractionExamples
- DisablePointersExample
- PrimaryPointerExample
- SolverExamples
- HandCoachExamples
- SlateExample (only uses slate / no overrides from prefab)
- RiggedHandVisualizer
- EyeTrackingDemo-03-Navigation

**Prefabs**:

- Cheese
- Slate (BoundingBox inherited in SlateUGUI and SlateUGUIHololens2Buttons but doesn't have overrides there)

### CODE REVIEW

File: Assets/MRTK/Core/Inspectors/Utilities/InspectorUIUtility.cs [hunk 1]
```diff
@@ -6,6 +6,7 @@
 using System.Linq;
 using System.Reflection;
 using UnityEditor;
+using UnityEditor.Experimental.SceneManagement;
```
Comment (julenka) [inline Assets/MRTK/Core/Inspectors/Utilities/InspectorUIUtility.cs:9]: why do we need to reference scenemanagement here? Seems unrelated.

File: Assets/MRTK/SDK/Features/UX/Scripts/BoundsControl/Visuals/RotationHandles.cs [hunk 2]
```diff
@@ -49,7 +49,7 @@ protected override Quaternion GetRotationRealignment(int handleIndex)
             }
             else // y axis 
             {
-                return Quaternion.FromToRotation(Vector3.forward, Vector3.up);
+                return Quaternion.identity;
```
Comment (julenka) [inline Assets/MRTK/SDK/Features/UX/Scripts/BoundsControl/Visuals/RotationHandles.cs:52]: This seems like a different behavior (no rotation vs rotation 90 degrees about x) -- is that intentional?

**Review by julenka — APPROVED** [review]
Did a quick skim -- overall looks good! There are quite a lot of changes to assets and prefabs, you might want to double check that these are all intentional.

Since this is going into 2.5, what sort of testing have you done to ensure this does not regress changes, or what testing will other team members need to do to verify these changes?

File: Assets/MRTK/SDK/Features/UX/Scripts/BoundsControl/Visuals/RotationHandles.cs (hunk 2 above)
Comment (thalbern) [reply Assets/MRTK/SDK/Features/UX/Scripts/BoundsControl/Visuals/RotationHandles.cs:52]: yes! this was a mistake carried over from translation handles. this always used to be identity. when i did the translation handles integration this sneaked in as a mistake and flattened rotation handles had the wrong rotation applied

File: Assets/MRTK/Core/Inspectors/Utilities/InspectorUIUtility.cs (hunk 1 above)
Comment (thalbern) [reply Assets/MRTK/Core/Inspectors/Utilities/InspectorUIUtility.cs:9]: it's for the PrefabStageUtility to do the prefab checks for nested scriptable assets.

**Comment by thalbern** [comment]
> Did a quick skim -- overall looks good! There are quite a lot of changes to assets and prefabs, you might want to double check that these are all intentional.
> 
> Since this is going into 2.5, what sort of testing have you done to ensure this does not regress changes, or what testing will other team members need to do to verify these changes?

I manually checked all the prefabs and scenes - they were not done by the tool as I couldn't trust the tool with the dependency problem we're having with it.  It's scenes and prefabs that still had references to old bounding box. 
There's 2 prefabs: Slate & Cheese - those are the most important to check IN scenes. So wherever those are used in any of the scenes we have to make sure everything looks like it should.

Scene list (scenes that are either using those prefabs or having their own bounding boxes in gameobject instances)
- EyeTrackingDemo-03-Navigation
- HandInteractionExamples
- DisablePointersExample
- PrimaryPointerExample
- SolverExamples
- HandCoachExamples
- SlateExample (only uses slate / no overrides from prefab)
- RiggedHandVisualizer

Prefabs:
- Cheese
- Slate (BoundingBox inherited in SlateUGUI and SlateUGUIHololens2Buttons but doesn't have overrides there)

Basically we have to check all those scenes and verify they behave and look the same as they used to. 
Note that the flattenAxisDisplayScale was not properly applied before, so that is an intended change. Also I've changed defaults in the Cheese prefab as the values were set to unusable numbers and every scene using the cheese was overriding those values anyways. I set proximity effect and handle size values to reasonable numbers now and that also fixes the primary pointer example scene and saves us some overrides in other scenes (makes it a bit cleaner)

**Review by CDiaz-MS — APPROVED** [review]
Tested out adding bounding box to my scene and upgrading.  

The button and deprecated warning for Bounding Box is displayed at the bottom of the component and I did not see it initially because I had to scroll.  It might be more obvious to users that the component is deprecated if we display the warning as the first section in the inspector instead of the last section.

**Comment by elbuhofantasma** [comment]
**Issue description**
A 'BoxColliders does not support negative scale or size' warning is issued when playing the scene with the in-editor player. There are 7 warnings generated when the scene is started for "SceneContent/BoundingBoxExample/HoloLens1 Style/Platonic/rigRoot/corner_0/visualsScale/visuals" with the '../corner_0/..' incrementing by one up to '../corner_7/..' in the Unity console. You are still able to interact with the scene as expected.

**Scenes effected**
HandInteractionExamples
DisabledPointers

**Unity version** 2019.4.9f1
**Platform** Standalone
**Devices connected** No devices were connected to the machine.

**Screenshots**
boxcollider
corner0
corner1
corner2
corner3
corner4
corner5
corner6

**Comment by elbuhofantasma** [comment]
**Issue description**
When playing the HandInteractionTouchablesExample scene with the in-editor player, two 'referenced scripts are missing' warnings are generated. 

**Unity version** 2019.4.9f1
**Platform** Standalone
**Devices connected** No devices were connected to the machine.

**Log output**
```
The referenced script (Unknown) on this Behaviour is missing! 
(Filename: C:\buildslave\unity\build\Runtime/Scripting/ManagedReference/SerializableManagedRef.cpp Line: 199)

The referenced script on this Behaviour (Game Object 'Plane') is missing! 
(Filename: C:\buildslave\unity\build\Runtime/Mono/ManagedMonoBehaviourRef.cpp Line: 334)
```

**Comment by elbuhofantasma** [comment]
**Issue description**
When using the Unity in-editor player to load the EyeTrackingDemo-03-Navigation scene, several errors are generated. 

**Unity version** 2019.4.9f1
**Platform** Standalone
**Devices connected** No devices were connected to the machine.

**PressableButton (2) NullReferenceException**
pressablebuttonnullreferenceexception

```
NullReferenceException: Object reference not set to an instance of an object
  at Microsoft.MixedReality.Toolkit.UI.Interactable.ResolveInputAction (System.Int32 index) [0x00001] in D:\unity\8508\MixedRealityToolkit-Unity\Assets\MRTK\SDK\Features\UX\Interactable\Scripts\Interactable.cs:1160 
  at Microsoft.MixedReality.Toolkit.UI.Interactable.Initialize () [0x0001f] in D:\unity\8508\MixedRealityToolkit-Unity\Assets\MRTK\SDK\Features\UX\Interactable\Scripts\Interactable.cs:758 
  at Microsoft.MixedReality.Toolkit.UI.Interactable.EnsureInitialized () [0x00016] in D:\unity\8508\MixedRealityToolkit-Unity\Assets\MRTK\SDK\Features\UX\Interactable\Scripts\Interactable.cs:743 
  at Microsoft.MixedReality.Toolkit.UI.Interactable.Awake () [0x00001] in D:\unity\8508\MixedRealityToolkit-Unity\Assets\MRTK\SDK\Features\UX\Interactable\Scripts\Interactable.cs:616 
 
(Filename: Assets/MRTK/SDK/Features/UX/Interactable/Scripts/Interactable.cs Line: 1160)
```

**ToggleProfileButton NullRefereceException**
toggleprofilebuttonnullreferenceexception

```
NullReferenceException: Object reference not set to an instance of an object
  at Microsoft.MixedReality.Toolkit.UI.Interactable.ResolveInputAction (System.Int32 index) [0x00001] in D:\unity\8508\MixedRealityToolkit-Unity\Assets\MRTK\SDK\Features\UX\Interactable\Scripts\Interactable.cs:1160 
  at Microsoft.MixedReality.Toolkit.UI.Interactable.Initialize () [0x0001f] in D:\unity\8508\MixedRealityToolkit-Unity\Assets\MRTK\SDK\Features\UX\Interactable\Scripts\Interactable.cs:758 
  at Microsoft.MixedReality.Toolkit.UI.Interactable.EnsureInitialized () [0x00016] in D:\unity\8508\MixedRealityToolkit-Unity\Assets\MRTK\SDK\Features\UX\Interactable\Scripts\Interactable.cs:743 
  at Microsoft.MixedReality.Toolkit.UI.Interactable.Awake () [0x00001] in D:\unity\8508\MixedRealityToolkit-Unity\Assets\MRTK\SDK\Features\UX\Interactable\Scripts\Interactable.cs:616 
 
(Filename: Assets/MRTK/SDK/Features/UX/Interactable/Scripts/Interactable.cs Line: 1160)
```

**Assertion failed**
contentassertionfailed

```
Loaded scene 'Temp/__Backupscenes/0.backup'
	Deserialize:            126.371 ms
	Integration:            112.949 ms
	Integration of assets:  0.080 ms
	Thread Wait Time:       11.856 ms
	Total Operation Time:   251.256 ms
Assertion failed
UnityEngine.DebugLogHandler:Internal_Log(LogType, LogOption, String, Object)
UnityEngine.DebugLogHandler:LogFormat(LogType, Object, String, Object[])
UnityEngine.Logger:Log(LogType, Object, Object)
UnityEngine.Debug:Assert(Boolean, Object)
Microsoft.MixedReality.Toolkit.Input.Utilities.CanvasUtility:Start() (at Assets\MRTK\Services\InputSystem\Utilities\CanvasUtility.cs:43)
 
(Filename: Assets/MRTK/Services/InputSystem/Utilities/CanvasUtility.cs Line: 43)
```

**No EventSystem detected**
noeventsystemdetected

```
No EventSystem detected. UI events will not be propagated to Unity UI.
UnityEngine.DebugLogHandler:Internal_Log(LogType, LogOption, String, Object)
UnityEngine.DebugLogHandler:LogFormat(LogType, Object, String, Object[])
UnityEngine.Logger:Log(LogType, Object)
UnityEngine.Debug:LogError(Object)
Microsoft.MixedReality.Toolkit.Input.Utilities.CanvasUtility:Start() (at Assets\MRTK\Services\InputSystem\Utilities\CanvasUtility.cs:48)
 
(Filename: Assets/MRTK/Services/InputSystem/Utilities/CanvasUtility.cs Line: 48)
```

**EyeTrackingDemos ManagerComponents**
eytracjubgnabagercompnullreferenceexception

```
NullReferenceException: Object reference not set to an instance of an object
  at Microsoft.MixedReality.Toolkit.Examples.Demos.EyeTracking.LoadProfilesOnStartup.Start () [0x00001] in D:\unity\8508\MixedRealityToolkit-Unity\Assets\MRTK\Examples\Demos\EyeTracking\General\Scripts\Utils\LoadProfilesOnStartup.cs:20 
 
(Filename: Assets/MRTK/Examples/Demos/EyeTracking/General/Scripts/Utils/LoadProfilesOnStartup.cs Line: 20)

```

**Comment by thalbern** [comment]
> Tested out adding bounding box to my scene and upgrading.
> 
> The button and deprecated warning for Bounding Box is displayed at the bottom of the component and I did not see it initially because I had to scroll. It might be more obvious to users that the component is deprecated if we display the warning as the first section in the inspector instead of the last section.

good idea. i'll move it up!

**Comment by thalbern** [comment]
Talked with @elbuhofantasma  - eye tracking scene problems and missing references in scenes are going to go into a separate issue as they're unrelated to this change.

Fixing collider warnings now in this PR

**Comment by thalbern** [comment]
fixing warning in follow up PR - will link after opening

**Comment by thalbern** [comment]
> **Issue description**
> A 'BoxColliders does not support negative scale or size' warning is issued when playing the scene with the in-editor player. There are 7 warnings generated when the scene is started for "SceneContent/BoundingBoxExample/HoloLens1 Style/Platonic/rigRoot/corner_0/visualsScale/visuals" with the '../corner_0/..' incrementing by one up to '../corner_7/..' in the Unity console. You are still able to interact with the scene as expected.
> 
> **Scenes effected**
> HandInteractionExamples
> DisabledPointers
> 
> **Unity version** 2019.4.9f1
> **Platform** Standalone
> **Devices connected** No devices were connected to the machine.
> 
> **Screenshots**
> boxcollider
> corner0
> corner1
> corner2
> corner3
> corner4
> corner5
> corner6

opened follow up PR here https://github.com/microsoft/MixedRealityToolkit-Unity/pull/8561
